### ISSUE TITLE
So far so good!

### ISSUE THREAD

**Issue by hollapm** [quielb/hass-airscape#1]
I wanted to let you know that I added the airscape to custom_components, rebooted, added the information to configuration.yaml for 2 fans, and I am now able to control both fans within Home Assistant.  

Ultimately I want to program the fan so that when the weather is cooler outside (and windows are open) to turn on the fan.  My coding also looks at humidity, outside vs attic vs inside temp, if smoke alarms are going off, turn off if min temp is reached, etc.

I created a WebCore instance using smartthings integration of airscape and it works well, but prefer to move to Home Assistant.

Would love to figure out how to import or re-create Webcore into HA.

**Comment by quielb** [comment]
Here is the automation I use that may help speed up your process

```yaml
- alias: WHF Auto Start
  trigger:
    - platform: state
      entity_id:
        - sensor.inside_temperature_max
        - sensor.outside_temperature
    - platform: state
      entity_id: binary_sensor.whf_airflow
      to: 'off'
  condition:
    - condition: template
      value_template: "{{ (states('sensor.inside_temperature_max')|float) > (states('sensor.outside_temperature')|float) }}"
    - condition: state
      entity_id: binary_sensor.whf_airflow
      state: 'off'
    - condition: state
      entity_id: fan.whole_house
      state: 'off'
    - condition: time
      after: '19:00:00'
      before: '05:59:00'
    - condition: numeric_state
      entity_id: sensor.purpleair
      below: 100
    - condition: numeric_state
      entity_id: sensor.inside_temperature_avg
      above: input_number.whf_low_temp
  action:
    - service: fan.turn_on
      data:
        entity_id: fan.whole_house

- alias: WHF Auto Shut Off
  trigger:
    - platform: numeric_state
      entity_id: sensor.inside_temp_slope
      above: .001
      for:
        minutes: 20
    - platform: state
      entity_id: binary_sensor.whf_airflow
    - platform: numeric_state
      entity_id: sensor.purpleair
      above: 100
    - platform: state
      entity_id: sensor.inside_temperature_avg
  condition:
    - condition: state
      entity_id: fan.whole_house
      state: 'on'
  action:
    - choose:
        - conditions:
            - condition: time
              after: '06:00:00'
              before: '11:00:00'
            - condition: numeric_state
              entity_id: sensor.inside_temp_slope
              above: 0.05
          sequence:
            - service: fan.turn_off
              data:
                entity_id: fan.whole_house
        - conditions:
            - condition: or
              conditions:
                - condition: state
                  entity_id: binary_sensor.whf_airflow
                  state: "on"
                - condition: numeric_state
                  entity_id: sensor.purpleair
                  above: 100
                - condition: numeric_state
                  entity_id: sensor.inside_temperature_avg
                  below: input_number.whf_low_temp
          sequence:
            - service: fan.turn_off
              data:
                entity_id: fan.whole_house

- alias: WHF Auto Speed Control
  trigger:
    platform: time_pattern
    minutes: "/30"
  condition:
    - condition: state
      entity_id: fan.whole_house
      state: 'on'
    - condition: state
      entity_id: input_select.home_mode
      state: 'Sleep'
  action:
    - service_template: >-
        {% if (
                (states('sensor.inside_temperature_avg')|float) -
                (states('sensor.outside_temperature')|float)
              ) > 5 -%}
          fan.airscape_speed_up
        {% else -%}
          fan.airscape_slow_down
        {%- endif %}
      data:
        entity_id: fan.whole_house
```

**Comment by hollapm** [comment]
This is great.  I'm starting to configure mine utilizing the work you did.  Couple of questions around the sensor entities, and what are they.
sensor.inside_temperature_max - is this the max temp that you want within the home, IE no. warmer than 78 degrees?  Or is it taking the temperature sensors in the home and taking the max value collected over 24 hrs?

binary_sensor.whf_airflow - where is it getting the airflow value from?  I know that the airscape fans are able to calculate temperature, speed, airflow etc, but I don't believe the fan control is able to pull this data?

sensor.purpleair - what is this? ;-)

sensor.inside_temperature_avg - is this taking the average temp over 24hrs, or is it the temp you'd ultimately like the house to be at?

sensor.inside_temp_slope - I'm just learing about slope, so don't know much about this... ;-(

This is quite helpful.  I'm hoping I can integrate some of the code I did from WebCore into Home Assistant.  What you have is a huge boost for me ;-)

**Comment by quielb** [comment]
> sensor.inside_temperature_max

I have several indoor temperature sensors.  Basically I'm looking for the hottest part of the house to see if it higher than the outside temperature.  This is the realtime max value of that group of sensors.  It's not averaged over a time peroid.
```yaml
- platform: min_max
  name: Inside Temperature MAX
  type: max
  entity_ids:
    - sensor.master_temperature
    - sensor.nook_temperature
    - sensor.my_ecobee_temperature
    - sensor.kids_room_temperature
```

> binary_sensor.whf_airflow

I have added an attribute to identify doors/windows that I would open to run the whole house fan.  The binary sensor is a `safety` device class sensor that indicates if it is safe to run the whole house fan (I have enough windows/doors open)
```yaml
- platform: template
  sensors:
    whf_airflow:
      friendly_name: "Open Windows for WHF"
      device_class: safety
      value_template: >-
        {{ (states.binary_sensor
          | selectattr('attributes.whf_airflow')
          | selectattr('state', 'eq', 'on')
          | list
          | count) < 2
        }}
```

> sensor.purpleair

This is an outdoor AQI sensor. We live in an area that has a lot of wild fires.  I don't want to pull all that bad air into the house so I shut off the WHF if the air outside is bad.

> sensor.inside_temperature_avg

This is just like the max_inside but this time its the average of all the sensors.  I compare the average inside temperature to shutoff temp to see if its cool enough throughout the house to turn off the fan, and prevent it from restarting.

> sensor.inside_temperature_avg

I found that comparing the inside temp to the outside temp to shut off the fan wasn't working well.  Because of the way my house is oriented I get the morning sun of the front side of my house.  The WHF will pull in that warm air because it is still "cooler" outside than it is inside according to the sensors.  An outdoor temp sensor doesn't account for the heat from solar radiation, which is the heat that gets pulled in during the morning hours as the sun beats on the house.  I take all of inside temp sensors and feed them in to influxdb.  Then I use the influxdb component to pull back the slope value of those sensors.  The slope of a graph indicates direction of the graph.  If the slope is positive that means the line is going up, and the opposite is true for a negative slope.  So in may case if the slope is positive that means the temperature is going up, not down.  Since I don't want the heat from outside to increase the temp inside I shut down the fan based on slope and not just comparing inside vs outside temperature.

**Comment by hollapm** [comment]
This makes sense!  
Dumb question, how did you add an attribute (whf_airflow) to the doors/windows you want to monitor?  I have 2 fans - one for the master and another for the whole house, so want to add an attribute called whf_master_airflow and another called whf_main_airflow.  I feel this should be easy but am not having success.

I created some sensors based off of groups:
`     door_count_master:
        value_template: >-
           {{ states | selectattr('entity_id','in',state_attr('group.master_doors','entity_id')) | selectattr('state','eq','on') | list | count | int}}
      window_count_master:
        value_template: >-
           {{ states | selectattr('entity_id','in',state_attr('group.master_windows','entity_id')) | selectattr('state','eq','on') | list | count | int}}
      door_count_main:
        value_template: >-
           {{ states | selectattr('entity_id','in',state_attr('group.main_doors','entity_id')) | selectattr('state','eq','on') | list | count | int}}
      window_count_main:
        value_template: >-
           {{ states | selectattr('entity_id','in',state_attr('group.main_windows','entity_id')) | selectattr('state','eq','on') | list | count | int}}
 `
which is giving me a count, but am not certain how to create the binary sensor which includes the proper attributes so that I can reference windows/doors that are open/safe to turn on the fan.  

Also is it possible to have it so that is say 2 windows are open that you only have the fan max at speed 3, but if 5 windows are open it can max at speed 7 and if over 5 then speed 10?

Same with if my smoke detector is going off, for it to immediately turn off the WHF.  I think this one would be easy, as it would be included in the automation, auto shut off...

Im new to the way the coding is set up in home assistant - more familiar with the WebCore side but am starting to see some of the similarities.

**Comment by quielb** [comment]
There are 2 ways to add attributes. The "old school" way is to use customize.yaml.  There is also an option in the UI on the configuration page.  On the Customizations page use the "pick an attribute to override" drop down and select other.  That will allow you to add custom attributes.

> I created some sensors based off of groups which is giving me a count...

A binary sensor is a true/false entity.  While having counts of your objects is useful for your other request, the template has to return true/false for a binary sensor.  By adding a comparison, something like less than or greater than a number it becomes a true false statement and can be used as a binary sensor.

>  Also is it possible to have it so that is say 2 windows are open that you only have the fan max at speed 3, but if 5 windows are open it can max at speed 7 and if over 5 then speed 10?

Sure. Just create an automation that triggers on the state of your count object.  Then you can use the "choose" primitive as a sort of switch statement.

> Same with if my smoke detector is going off, for it to immediately turn off the WHF. I think this one would be easy, as it would be included in the automation, auto shut off...

Correct

**Comment by hollapm** [comment]
Sweet, two more questions ;-)
For the slope, can you share what you used in influxdb?  I have never used it and just started to play with it.  For the slope, I assume that you have one side of the house listed first, then the other last hence how the slope is created due to temp change?

The second question is the Sleep
    - condition: state
      entity_id: input_select.home_mode
      state: 'Sleep'

Is this used so that the fan only operates when you are home and home_mode "sleep" is called?  I assume this is an automation also created?  do you mind sharing ;-)

So far it's working quite well.  I need to likely add humidity to mine, as it was starting to get humid in the house (it was cooler outside than in, so would likely compare outside humiditiy to inside, and once inside humidity gets above X to turn off.

**Comment by hollapm** [comment]
One more question, is it possible to pull the data that the fan has, IE attic temp. CFM, power etc?
Would love to know the CFM and attic temp.  This way if the attic is really hot, then the WHF could run to cool the attic down fast then turn off.  Or use the CFM to decipher how many windows I need opened.

Let me know how I can help!

**Comment by quielb** [comment]
> Sweet, two more questions ;-)
> For the slope, can you share what you used in influxdb? I have never used it and just started to play with it. For the slope, I assume that you have one side of the house listed first, then the other last hence how the slope is created due to temp change?

Slope is an indicator of change over time. Think about the average inside temperature on a graph. The slope tells if the temperature is going up or down. Slope is positive the graph is going up. Slope is negative the temperature is going down. Here is the influxdb query I use:
```yaml
- platform: influxdb
  queries:
    - name: inside temp slope
      unit_of_measurement: °F
      value_template: '{{ value | round(2) }}'
      measurement: '(select derivative(mean(value)) as value from "°F" where "entity_id" = ''inside_temperature_avg'' AND time > now() - 1h group by time(5m) fill(none))'
      where: 'time > now() - 1h'
      field: 'value'
      group_function: last
```
Basically the query limits data points to the last hour.  Then takes those data points and creates a 5 minute rolling average (bucket average). Then we find the slope (derivative) of that data. Then we only return the last data point. The influxdb sensor complains if you return more than 1 data point. If you have influxdb running, I would suggest getting graphana running so that you can visualize what this function does.  It might help. 

> The second question is the Sleep
> - condition: state
> entity_id: input_select.home_mode
> state: 'Sleep'
> 
> Is this used so that the fan only operates when you are home and home_mode "sleep" is called? I assume this is an automation also created? do you mind sharing ;-)
> 

My house has 3 states: Home, Away, Sleep. And certain things happen when a house changes into one of those states. Lights turn on or off, temperate on the thermostat gets set, doors lock, garage door closes, etc. I limit the automatic speed control of the WHF to only work when the house is in the sleep state. Because of the location of the fan in the house it can get loud when running at full speed.  It makes it hard to watch TV. But when I'm sleeping the noise of the fan is just white noise and really doesn't bother me. I've found that using my fan I can only get it down to within about 5 degrees of the outside temperature. So it doesn't make sense to speed it up to max when its not the coolest part of the night.  
```yaml
- alias: WHF Auto Speed Control
  trigger:
    platform: time_pattern
    minutes: "/30"
  condition:
    - condition: state
      entity_id: fan.whole_house
      state: 'on'
    - condition: state
      entity_id: input_select.home_mode
      state: 'Sleep'
  action:
    - service_template: >-
        {% if (
                (states('sensor.inside_temperature_avg')|float) -
                (states('sensor.outside_temperature')|float)
              ) > 5 -%}
          fan.airscape_speed_up
        {% else -%}
          fan.airscape_slow_down
        {%- endif %}
      data:
        entity_id: fan.whole_house
```

> One more question, is it possible to pull the data that the fan has, IE attic temp. CFM, power etc?
Would love to know the CFM and attic temp. This way if the attic is really hot, then the WHF could run to cool the attic down fast then turn off. Or use the CFM to decipher how many windows I need opened.

It's possible to collect that data. I can add it as attributes to the entity. But making a determination of speed based on windows open isn't the best way to use a WHF. As long as you have 1 window open, you're not going to burn out the fan. And using it based on attic temperature won't give you the results you are looking for. It is very common for the attic to be hotter than outside. In that case it is also common for it to be hotter outside than inside. If you were to turn the fan on to cool the attic, you would accomplish that.  But you would also warm up the inside of your house as well. The basic principle of a WHF is deltaT between the inside and outside temperatures, The attic is a bystander in all of this. It gets cooled because of the airflow from outside->inside->attic. It you want to cool the attic during the day either increase the net free air of ventilation to get better convection cooling, or get something like a solar powered gable fan.

**Comment by hollapm** [comment]
I'm new to the programming side (my old way was Pascal and some C+, as well as if/then statements...)
I'm having a challenge getting the slope to work for me, and now know why you want it as it runs longer in the morning and starts to pull in warmer air.

this is what I have for my slope and it's not creating an entity in Home assistant.  Not certain what I have wrong, IE formatting or what:
`# Influx Slope
  - platform: influxdb
    queries:
      - name: insidetempslope
        unit_of_measurement: °F
        value_template: '{{ value | round(2) }}'
        measurement: '(select derivative(mean(value)) as value from "°F" where "entity_id" = ''kmpcil_res001_526d0c1a_temperature'' AND time > now() - 1h group by time(5m) fill(none))'
        where: 'time > now() - 1h'
        field: 'value'
        group_function: last`

I'm able to see the sensor in influxdb, I'm able to also see it in grafana.  But when I add the above into my sensor for home assistant its not creating a sensor called insidetempslope.  I am able to create a trend using the trend feature, but it's not nearly that stable for me yet.

I'm wondering if I need to do an average for my one sensor (which will flatten out the curve some?)
Any pointers?

**Comment by hollapm** [comment]
looks like I'm having issues with influxDB.  Partially fixed but not how I want it.  Playing with the slope now.

**Comment by hollapm** [comment]
InfluxDB issue is now resolved, but it doesn't appear to be running the query properly.  The sensor is created but the slope never changes.  

```
  - platform: influxdb
    host: 192.168.86.75
    port: 8086
    username: hass
    password: XXXX
    queries:
      - name: inside_temp_slope_master
        unit_of_measurement: °F
        value_template: '{{ value | round(2) }}'
        measurement: '(select derivative(mean(value)) as value from "°F" where "entity_id" = ''sensor.inside_temperature'' AND time > now() - 1h group by time(5m) fill(none))'
        where: 'time > now() - 1h'
        field: 'value'
        group_function: last
        database: home_assistant
```

Any thoughts on what may be off?

**Comment by hollapm** [comment]
We're you able to pull any additional stats from the fan like temp, cfm, power, time remaining etc? Temp would be most important although I received an error and would love for it to pull this up.

Reduced by Pressure Control
Pressure Control means that the differential pressure switch has determined that your house pressure is too low. Open more windows and reset by clicking the fan-speed-up button

**Comment by quielb** [comment]
Yes I still have plans to add all the attributes. I am currently working on rebuilding my home environment which includes my development boxes. As soon as I get through all that, I will get back to work on this.

**Comment by quielb** [comment]
I've added all the fan data as attributes to the fan entity. If you want to make sensors out of them, then just create template sensors based on the attributes.

I have also added the necessary pieces to be able to add the component via HACS. For now it is a custom repository and not included in the default store. You will have to add the custom repository to your HACS config to be able to mange the install/update of the custom component.

**Comment by hollapm** [comment]
Fantastic,I've updated the custom repository and will play with the settings later this week!  Thank you for all the work!    On Saturday, February 13, 2021, 06:05:32 PM PST, Barry Quiel <notifications@github.com> wrote:  
 
 


I've added all the fan data as attributes to the fan entity. If you want to make sensors out of them, then just create template sensors based on the attributes.

I have also added the necessary pieces to be able to add the component via HACS. For now it is a custom repository and not included in the default store. You will have to add the custom repository to your HACS config to be able to mange the install/update of the custom component.

**Comment by hollapm** [comment]
It works perfectly!  Thank you!!!
    On Tuesday, February 16, 2021, 05:49:31 PM PST, Paul Holland <hollapm@yahoo.com> wrote:  
 
  
Fantastic,I've updated the custom repository and will play with the settings later this week!  Thank you for all the work!    On Saturday, February 13, 2021, 06:05:32 PM PST, Barry Quiel <notifications@github.com> wrote:  
 
 


I've added all the fan data as attributes to the fan entity. If you want to make sensors out of them, then just create template sensors based on the attributes.

I have also added the necessary pieces to be able to add the component via HACS. For now it is a custom repository and not included in the default store. You will have to add the custom repository to your HACS config to be able to mange the install/update of the custom component.
